mediatimestamp/immutable.py

```diff
@@ -969,6 +987,147 @@ def intersect_with(self, tr):
+
+    def ends_later_than_timerange(self, other):
+        """Returns true if this timerange ends later than the end of the other."""
+        return (other.bounded_after() and
+                (not self.bounded_after() or
+                 (self.end > other.end or
+                  (self.end == other.end and
+                   self.includes_end() and
+                   not other.includes_end()))))
+
+    def overlaps_with_timerange(self, other):
+        """Returns true if this timerange and the other overlap."""
+        return (not self.is_earlier_than_timerange(other) and not self.is_later_than_timerange(other))
+
+    def is_contiguous_with_timerange(self, other):
+        """Returns true if the union of this timerange and the other would be a valid timerange"""
+        return (self.overlaps_with_timerange(other) or
+                (self.is_earlier_than_timerange(other) and
+                 self.end == other.start and
+                 (self.includes_end() or other.includes_start())) or
+                (self.is_later_than_timerange(other) and
+                 self.start == other.end and
+                 (self.includes_start() or other.includes_end())))
+
+    def union_with_timerange(self, other):
+        """Returns the union of this timerange and the other.
+        :raises: ValueError if the ranges are not contiguous."""
+        if not self.is_contiguous_with_timerange(other):
+            raise ValueError("Timeranges {} and {} are not contiguous, so cannot take the union.".format(self, other))
+
+        inclusivity = TimeRange.EXCLUSIVE
+        if self.start == other.start:
+            start = self.start
+            inclusivity |= ((self.inclusivity | other.inclusivity) & TimeRange.INCLUDE_START)
+        elif self.starts_earlier_than_timerange(other):
+            start = self.start
+            inclusivity |= (self.inclusivity & TimeRange.INCLUDE_START)
+        else:
+            start = other.start
+            inclusivity |= (other.inclusivity & TimeRange.INCLUDE_START)
+
+        if self.end == other.end:
+            end = self.end
+            inclusivity |= ((self.inclusivity | other.inclusivity) & TimeRange.INCLUDE_END)
+        elif self.ends_later_than_timerange(other):
+            end = self.end
+            inclusivity |= (self.inclusivity & TimeRange.INCLUDE_END)
+        else:
+            end = other.end
+            inclusivity |= (other.inclusivity & TimeRange.INCLUDE_END)
+
+        return TimeRange(start, end, inclusivity)
+
+    def split_at(self, timestamp):
+        """Splits a timerange at a specified timestamp.
+
+        :param timestamp: the timestamp to split at
+        :returns: A pair of TimeRange objects
+        :raises: ValueError if timestamp not in self"""
+
+        if timestamp not in self:
+            raise ValueError("Cannot split range {} at {}".format(self, timestamp))
+
+        return (TimeRange(self.start, timestamp, (self.inclusivity & TimeRange.INCLUDE_START)),
```

Comment:
Should a special case be made for single timestamp time range and perhaps return ([ts, ts], None)?
If self is [ts, ts], i.e. a single timestamp, and a split is made on ts then this results in ([ts, ts), [ts, ts]). The first element would be invalid.

Reply:
@philipnbbc I have elected to improve the documentation and add testing to make intention clear there. The range [ts, ts) is not invalid, but it does evaluate as equal to TimeRange.never().

The intention is that of the returned ranges the split point will always be located in the later one.

Reply:
OK, thanks. There is a slightly confusing typo in the update: `*seconds*` should be `*second*`. Otherwise still LGTM,